apps/prairielearn/src/pages/enroll/enroll.sql

```diff
@@ -44,23 +44,93 @@ ORDER BY
   ci.id DESC;
 
 -- BLOCK enroll
-INSERT INTO
-  enrollments AS e (user_id, course_instance_id)
+WITH
+  enrollment_limits AS (
+    SELECT
+      i.id AS institution_id,
+      i.yearly_enrollment_limit AS institution_yearly_enrollment_limit,
+      -- Course instance enrollment limit overrides institution course instance
+      -- enrollment limit to allow for higher limits on individual course instances.
+      COALESCE(
+        ci.enrollment_limit,
+        i.course_instance_enrollment_limit
+      ) AS course_instance_enrollment_limit
+    FROM
+      institutions AS i
+      JOIN pl_courses AS c ON (c.institution_id = i.id)
+      JOIN course_instances AS ci ON (ci.course_id = c.id)
+    WHERE
+      ci.id = $course_instance_id
+      -- Lock institution to prevent concurrent enrollments that would exceed
+      -- the enrollment limit.
+    FOR UPDATE OF
```

Comment:
NO KEY UPDATE to prevent blocking SELECTs?

Reply:
Yes!! Good call. I need to get in the habit of doing that.